tests/test_edgeql_casts.py

```diff
@@ -2204,6 +2204,112 @@ async def test_edgeql_casts_json_11(self):
+                to_json({'{"a": 3000, "b": -1}', '{"a": 1, "b": 12}'});
+            """,
+            [{"a": 3000, "b": -1}, {"a": 1, "b": 12}],
+        )
+
+        await self.assert_query_result(
+            r"""
+                SELECT <tuple<int64, int64>>
+                to_json({'[3000, -1]', '[1, 12]'})
+            """,
+            [[3000, -1], [1, 12]],
+        )
+
+        self.assertEqual(
+            await self.con.query(
+                r"""
+                    SELECT <tuple<int64, int64>>
+                    to_json({'[3000, -1]', '[1, 12]'})
+                """
+            ),
+            [(3000, -1), (1, 12)],
+        )
+
+        self.assertEqual(
+            await self.con.query(
+                r"""
+                    SELECT <tuple<json, json>>
+                    to_json({'[3000, -1]', '[1, 12]'})
+                """
+            ),
+            [('3000', '-1'), ('1', '12')],
+        )
+
+        self.assertEqual(
+            await self.con.query(
+                r"""
+                    SELECT <tuple<json, json>>
+                    to_json({'[3000, -1]', '[1, null]'})
+                """
+            ),
+            [('3000', '-1'), ('1', 'null')],
+        )
+
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'expected json number, null; got json string'):
+            await self.con.query(
+                r"""
+                    SELECT <tuple<a: int64, b: int64>>
+                    to_json('{"a": 1, "b": "2"}')
+                """
+            )
+
+        # This isn't really the best error message for this.
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'invalid null value in cast'):
+            await self.con.query(
+                r"""SELECT <tuple<a: int64, b: int64>>to_json('{"a": 1}')"""
+            )
+
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'invalid null value in cast'):
+            await self.con.query(
+                r"""SELECT <tuple<int64, int64>>to_json('[3000]')"""
+            )
+
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'invalid null value in cast'):
+            await self.con.query(
+                r"""
+                    SELECT <tuple<a: int64, b: int64>>
+                    to_json('[3000, 1000]')
+                """
+            )
+
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'invalid null value in cast'):
+            await self.con.query(
+                r"""SELECT <tuple<a: int64, b: int64>> to_json('"test"')"""
+            )
+
+        with self.assertRaisesRegex(
+                edgedb.InvalidValueError,
+                r'invalid null value in cast'):
+            await self.con.query(
+                r"""SELECT <tuple<json, json>> to_json('[3000]')"""
+            )
+
```

Comment:
Would a `<tuple<tuple<int64, str>, tuple<str, int64>> some_json` work (nested tuples)? What about `<tuple<array<tuple ... >>>` types?

Reply:
That ought to work. I'll add some tests.